packages/toolpad-app/src/toolpad/AppEditor/index.tsx

```diff
@@ -43,48 +42,73 @@ interface FileEditorProps {
 }
 
 function FileEditor({ appId }: FileEditorProps) {
-  const { dom, currentView } = useDom();
-
-  const app = appDom.getApp(dom);
-  const { pages = [] } = appDom.getChildNodes(dom, app);
+  const { currentView } = useDom();
 
+  const location = useLocation();
   const navigate = useNavigate();
 
-  const firstPage = pages.length > 0 ? pages[0] : null;
-
   React.useEffect(() => {
-    switch (currentView.kind) {
-      case 'page':
-        navigate(`/app/${appId}/pages/${currentView.nodeId || firstPage?.id}`, { replace: true });
-        break;
-      case 'connection':
-        navigate(`/app/${appId}/connections/${currentView.nodeId}`, { replace: true });
-        break;
-      case 'codeComponent':
-        navigate(`/app/${appId}/codeComponents/${currentView.nodeId}`);
-        break;
-      default:
+    if (currentView.kind === 'page') {
+      const newPathname = `/app/${appId}/pages/${currentView.nodeId}`;
+
+      if (newPathname !== location.pathname) {
+        navigate(
+          {
+            pathname: newPathname,
+          },
+          {
+            replace: true,
+          },
+        );
+      }
+    }
+
+    if (currentView.kind === 'connection') {
+      const newPathname = `/app/${appId}/connections/${currentView.nodeId}`;
+
+      if (newPathname !== location.pathname) {
+        navigate(
+          {
+            pathname: newPathname,
+          },
+          {
+            replace: true,
+          },
+        );
+      }
+    }
+
+    if (currentView.kind === 'codeComponent') {
+      const newPathname = `/app/${appId}/codeComponents/${currentView.nodeId}`;
+
+      if (newPathname !== location.pathname) {
+        navigate(
+          {
+            pathname: newPathname,
+          },
+          {
+            replace: true,
+          },
+        );
+      }
     }
-  }, [appId, currentView.kind, currentView.nodeId, firstPage?.id, navigate]);
+  }, [appId, currentView.kind, currentView.nodeId, location.pathname, navigate]);
 
   return (
-    <Routes>
-      <Route element={<AppEditorShell appId={appId} />}>
-        <Route path="connections/:nodeId" element={<ConnectionEditor appId={appId} />} />
-        <Route path="pages/:nodeId" element={<PageEditor appId={appId} />} />
-        <Route path="codeComponents/:nodeId" element={<CodeComponentEditor appId={appId} />} />
-        <Route
-          index
-          element={
-            firstPage ? (
-              <Navigate to={`pages/${firstPage.id}`} replace />
-            ) : (
-              <NoPageFound appId={appId} />
-            )
-          }
-        />
-      </Route>
-    </Routes>
+    <AppEditorShell appId={appId}>
+      {currentView.kind === 'page' ? (
```

Comment:
The intent of this isn't very clear from structuring the code this way. (The intent being "only one of these needs to be rendered"). Moreover, the last element needs to always be kept in sync with the lines above. Ideally this would be written as a `switch`. Infortunately there's still no way to write switch expressions or do pattern matching in javascript, but perhaps this could be written as:

```tsx
{(() => {
  switch (currentView.kind) {
    case 'page': return <PageEditor appId={appId} nodeId={currentView.nodeId} />;
    // ...
  }
})()}
```

It's a bit ugly with all those brackets. Alternatively this could be written in a `useMemo`, or extracted as a function.